plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
 	}
 
-	if !strings.Contains(firstCPU.flags, "aperfmperf") {
-		p.cpuBusyCycles = false
-		p.cpuBusyFrequency = false
-		p.cpuC0StateResidency = false
-		p.cpuC1StateResidency = false
+	if !slices.Contains(firstCPU.Flags, "aperfmperf") {
+		p.disableCPUMetric(cpuC0StateResidency)
+		p.disableCPUMetric(cpuC1StateResidency)
+		p.disableCPUMetric(cpuBusyCycles)
+		p.disableCPUMetric(cpuBusyFrequency)
 	}
 
-	if !strings.Contains(firstCPU.flags, "dts") {
-		p.cpuTemperature = false
+	if !slices.Contains(firstCPU.Flags, "dts") {
+		p.disableCPUMetric(cpuTemperature)
 	}
 
 	return nil
 }
 
-func contains[T comparable](s []T, e T) bool {
-	for _, v := range s {
-		if v == e {
-			return true
-		}
-	}
-	return false
-}
+// disableCPUMetric removes given cpu metric from cpu_metrics.
+func (p *PowerStat) disableCPUMetric(metricToDisable cpuMetricType) {
+	startLen := len(p.CPUMetrics)
+	p.CPUMetrics = slices.DeleteFunc(p.CPUMetrics, func(cpuMetric string) bool {
+		return cpuMetric == metricToDisable.String()
+	})
 
-func (p *PowerStat) areCoreMetricsEnabled() bool {
-	return p.msr != nil && len(p.msr.getCPUCoresData()) > 0
+	if len(p.CPUMetrics) < startLen {
+		p.Log.Warnf("%q is not supported by CPU, metric will not be gathered.", metricToDisable)
+	}
 }
 
-func (p *PowerStat) areGlobalMetricsEnabled() bool {
-	return p.rapl != nil
-}
+// disablePackageMetric removes given package metric from package_metrics.
+func (p *PowerStat) disablePackageMetric(metricToDisable packageMetricType) {
+	startLen := len(p.PackageMetrics)
+	p.PackageMetrics = slices.DeleteFunc(p.PackageMetrics, func(packageMetric string) bool {
+		return packageMetric == metricToDisable.String()
+	})
 
-func (p *PowerStat) GetCPUIDFromSocketID(socketID string) (string, error) {
-	for _, v := range p.cpuInfo {
-		if v.physicalID == socketID {
-			return v.cpuID, nil
-		}
+	if len(p.PackageMetrics) < startLen {
+		p.Log.Warnf("%q is not supported by CPU, metric will not be gathered.", metricToDisable)
 	}
-	return "", fmt.Errorf("can't find cpuID for socketID: %s", socketID)
 }
 
-// newPowerStat creates and returns PowerStat struct
-func newPowerStat(fs fileService) *PowerStat {
-	p := &PowerStat{
-		skipFirstIteration: true,
-		fs:                 fs,
-		logOnce:            make(map[string]error),
+// logErrorOnce takes an accumulator, a key string value error map, a key string and an error. It adds the error to the accumulator only if the
+// key is not in the logOnceMap. Additionally, if the key is not in logOnceMap map, adds the key to it. This is to prevent excessive error messages
+// from flooding the accumulator.
+func logErrorOnce(acc telegraf.Accumulator, logOnceMap map[string]struct{}, key string, err error) {
+	if _, ok := logOnceMap[key]; !ok {
+		acc.AddError(err)
+		logOnceMap[key] = struct{}{}
 	}
-
-	return p
 }
 
 func init() {
 	inputs.Add("intel_powerstat", func() telegraf.Input {
-		return newPowerStat(newFileService())
+		return &PowerStat{
+			MsrReadTimeout: 0,
```

Comment:
I think this is already the default value, so you can just do 
```suggestion
		return &PowerStat{}
```

Reply:
Done as part of commit [e2de6f6b](https://github.com/influxdata/telegraf/pull/14363/commits/e2de6f6b2577fe4e9ec7b420132b2750fe7e2654).